IdentityCore/src/cache/MSIDTokenCache.m

```diff
@@ -24,50 +24,386 @@
+                                                error:error];
+    
+    if (!legacyTokens)
+    {
+        return nil;
+    }
+    
+    for (MSIDToken *token in legacyTokens)
+    {
+        if (token.tokenType == MSIDTokenTypeRefreshToken
+            && token.clientId == clientId)
+        {
+            [allRTs addObject:token];
+        }
+    }
+    
+    return allRTs;
+}
+
+- (BOOL)saveAdalAT:(MSIDToken *)adalAT
+          clientId:(NSString *)clientId
+              user:(MSIDUser *)user
+          resource:(NSString *)resource
+           context:(id<MSIDRequestContext>)context
+             error:(NSError **)error
+{
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyWithAuthority:adalAT.authority
+                                                        clientId:clientId
+                                                        resource:resource
+                                                             upn:user.upn];
+    
+    return [_dataSource setItem:adalAT
+                            key:key
+                     serializer:_keyedArchiverSerialize
+                        context:context
+                          error:error];
+}
+
+- (BOOL)saveMsalAT:(MSIDToken *)msalAT
+         authority:(NSURL *)authority
+          clientId:(NSString *)clientId
+              user:(MSIDUser *)user
+            scopes:(NSOrderedSet<NSString *> *)scopes
+           context:(id<MSIDRequestContext>)context
+             error:(NSError **)error
+{
+    //delete all cache entries with intersecting scopes
+    //this should not happen but we have this as a safe guard against multiple matches
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyForAccessTokenWithAuthority:authority
+                                                                      clientId:clientId
+                                                                        scopes:scopes
+                                                                        userId:user.userIdentifier];
+    
+    NSArray<MSIDToken *> *allTokens = [_dataSource itemsWithKey:key serializer:_jsonSerializer context:context error:error];
+    
+    if (!allTokens)
+    {
+        return NO;
+    }
+
+    for (MSIDToken *token in allTokens)
+    {
+        if (token.tokenType == MSIDTokenTypeAccessToken
+            && [token.authority msidIsEquivalentAuthority:msalAT.authority]
+            && [token.scopes intersectsOrderedSet:msalAT.scopes])
+        {
+            MSIDTokenCacheKey *keyToDelete = [MSIDTokenCacheKey keyForAccessTokenWithAuthority:token.authority
+                                                                                      clientId:token.clientId
+                                                                                        scopes:token.scopes
+                                                                                        userId:user.userIdentifier];
+            
+            if(![_dataSource removeItemsWithKey:keyToDelete context:context error:nil])
```

Comment:
can we handle this in the datasource itself?

Reply:
The whole checking and removing logic?

Reply:
Let's discuss offline, I'm not completely sure how we can improve this logic.